### PR TITLE
Support Scriptable System

### PR DESCRIPTION
**Pull request by Roms1383**
Ok so this adds support for `ScriptableSystem`.

### CODE REVIEW

**Comment by Roms1383** [comment]
I've tried various ways locally (haven't pushed to avoid spamming) nothing does seem to work except using RTTI methods like so:

<details><summary>click to see code sample</summary>

```rs
let game = call!("GetGameInstance"() -> GameInstance).unwrap();
let rtti = RttiSystem::get();
let cls = rtti.get_class(CName::new(GameInstance::NAME)).unwrap();
let static_method = cls
    .static_methods()
    .iter()
    .find(|x| x.as_function().name() == CName::new("GetAudioSystem"))
    .unwrap();
let inst: Ref<AudioSystem> = static_method.as_function().execute(None, (game,)).unwrap();
let cls = rtti.get_class(CName::new(AudioSystem::CLASS_NAME)).unwrap();
let method: &Method = cls.get_method(CName::new("Play")).ok().unwrap();
method
    .as_function()
    .execute::<_, ()>(
        inst.instance().map(AsRef::as_ref),
        (
            CName::new("ono_v_effort"),
            EntityId::default(),
            CName::default(),
        ),
    )
    .unwrap();
```

</details>

I also noticed that `get_class` or `get_type` returned memory where `unk8` is `0`, but I'm not sure if it's related or not.

**Comment by Roms1383** [comment]
still same error, but at least I fixed incorrect types.

**Comment by Roms1383** [comment]
it seems the vft is wrong, see [RED4ext.SDK](https://github.com/WopsS/RED4ext.SDK/blob/109d7fb0044f86b90d6ccc71f2c30c8a5b6ca208/include/RED4ext/GameEngine.hpp#L32-L45) but here's what I get in debugger:

Screenshot 2024-07-17 125740

`AK` typically be AudioKinetics stuff

**Comment by Roms1383** [comment]
now game crashes on when method is called, I bet I forget to `mem::forget` somewhere ^^

**Comment by Roms1383** [comment]
To summarize with numerous mistake for anyone interested:
- vft functions first param is always `this`
- careful when transmuting / casting (type must be preserved as closely as possible when going from one to another)
- `ScriptGameInstance` and `GameInstance` are 2 different (but related) types
- `ScriptGameInstance` is only used to pass to scripts

**Comment by Roms1383** [comment]
here's how you can test quickly:

```rs
fn test_play() {
    let rtti = RttiSystem::get();
    let class = rtti.get_class(CName::new(AudioSystem::CLASS_NAME)).unwrap();
    let engine = GameEngine::get();
    let game = engine.game_instance();
    let system = game
        .get_system(class.as_type())
        .cast::<AudioSystem>()
        .unwrap();
    system.play(CName::new("ono_v_pain_long"), Opt::Default, Opt::Default);
}
```

and of course:
```rs
use red4ext_rs::{
    types::{CName, EntityId, IScriptable, Method, Native, Opt, Ref, ScriptClass},
    RttiSystem,
};

#[repr(C)]
pub struct AudioSystem {
    pub base: IScriptable,
    pub _padding0: [u8; 0x3E0],
}

unsafe impl ScriptClass for AudioSystem {
    const CLASS_NAME: &'static str = "gameGameAudioSystem";
    type Kind = Native;
}

impl AsRef<IScriptable> for AudioSystem {
    #[inline]
    fn as_ref(&self) -> &IScriptable {
        &self.base
    }
}

#[allow(dead_code)]
pub trait GameAudioSystem {
    fn play(&self, event_name: CName, entity_id: Opt<EntityId>, emitter_name: Opt<CName>);
    fn stop(&self, event_name: CName, entity_id: Opt<EntityId>, emitter_name: Opt<CName>);
    fn switch(
        &self,
        switch_name: CName,
        switch_value: CName,
        entity_id: Opt<EntityId>,
        emitter_name: Opt<CName>,
    );
}

impl GameAudioSystem for Ref<AudioSystem> {
    fn play(&self, event_name: CName, entity_id: Opt<EntityId>, emitter_name: Opt<CName>) {
        let rtti = RttiSystem::get();
        let cls = rtti.get_class(CName::new(AudioSystem::CLASS_NAME)).unwrap();
        let method: &Method = cls.get_method(CName::new("Play")).ok().unwrap();
        method
            .as_function()
            .execute::<_, ()>(
                unsafe { self.instance() }.map(AsRef::as_ref),
                (
                    event_name,
                    entity_id.unwrap_or_default(),
                    emitter_name.unwrap_or_default(),
                ),
            )
            .unwrap();
    }

    fn stop(&self, event_name: CName, entity_id: Opt<EntityId>, emitter_name: Opt<CName>) {
        let rtti = RttiSystem::get();
        let cls = rtti.get_class(CName::new(AudioSystem::CLASS_NAME)).unwrap();
        let method: &Method = cls.get_method(CName::new("Stop")).ok().unwrap();
        method
            .as_function()
            .execute::<_, ()>(
                unsafe { self.instance() }.map(AsRef::as_ref),
                (
                    event_name,
                    entity_id.unwrap_or_default(),
                    emitter_name.unwrap_or_default(),
                ),
            )
            .unwrap();
    }

    fn switch(
        &self,
        switch_name: CName,
        switch_value: CName,
        entity_id: Opt<EntityId>,
        emitter_name: Opt<CName>,
    ) {
        let rtti = RttiSystem::get();
        let cls = rtti.get_class(CName::new(AudioSystem::CLASS_NAME)).unwrap();
        let method: &Method = cls.get_method(CName::new("Switch")).ok().unwrap();
        method
            .as_function()
            .execute::<_, ()>(
                unsafe { self.instance() }.map(AsRef::as_ref),
                (
                    switch_name,
                    switch_value,
                    entity_id.unwrap_or_default(),
                    emitter_name.unwrap_or_default(),
                ),
            )
            .unwrap();
    }
}
```

File: src/types/game_engine.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+unsafe impl ScriptClass for ScriptableSystem {
+    type Kind = Native;
+
+    const CLASS_NAME: &'static str = "gameScriptableSystem";
+}
+
+impl AsRef<IScriptable> for ScriptableSystem {
+    fn as_ref(&self) -> &IScriptable {
+        unsafe { mem::transmute(&self.0._base._base) }
```
Comment (jac3km4) [inline src/types/game_engine.rs:99]:
```suggestion
        &self.0._base._base
```
seems we shouldn't need transmute here

File: src/types/game_engine.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
+use std::mem;
+
+use super::{IScriptable, Native, Ref, ScriptClass, Type};
+use crate::raw::root::RED4ext as red;
+use crate::types::WeakRef;
+use crate::{NativeRepr, VoidPtr};
+
+#[derive(Default)]
+#[repr(transparent)]
+pub struct GameInstance(red::ScriptGameInstance);
```
Comment (jac3km4) [inline src/types/game_engine.rs:10]:
```suggestion
pub struct ScriptGameInstance(red::ScriptGameInstance);
```
I'd follow the red4ext names here

File: src/types/game_engine.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    }
+}
+
+unsafe impl NativeRepr for GameInstance {
+    const NAME: &'static str = "ScriptGameInstance";
+}
+
+#[derive(Default)]
+#[repr(transparent)]
+pub struct RedGameInstance(red::GameInstance);
```
Comment (jac3km4) [inline src/types/game_engine.rs:27]:
```suggestion
pub struct GameInstance(red::GameInstance);
```
as above

File: src/types/game_engine.rs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    _unk60: VoidPtr,
+    _unk68: VoidPtr,
+}
+
+#[repr(transparent)]
+pub struct GameEngine(red::CGameEngine);
+
+impl GameEngine {
+    pub fn get<'a>() -> &'a Self {
+        unsafe { mem::transmute(&*red::CGameEngine::Get()) }
```
Comment (jac3km4) [inline src/types/game_engine.rs:80]:
```suggestion
        unsafe { mem::transmute(red::CGameEngine::Get()) }
```

File: src/types/game_engine.rs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    pub fn game_instance(&self) -> &RedGameInstance {
+        unsafe { mem::transmute(&*(*self.0.framework).gameInstance) }
```
Comment (jac3km4) [inline src/types/game_engine.rs:84]:
```suggestion
        unsafe { mem::transmute((*self.0.framework).gameInstance) }
```

File: src/types/game_engine.rs (hunk 1 above)
Comment (Roms1383) [reply src/types/game_engine.rs:99]: `IScriptable` and `red::IScriptable` are 2 different types, so it raises a compilation error without `transmute`.

Comment (jac3km4) [reply src/types/game_engine.rs:99]: oh okay

File: src/types/game_engine.rs (hunk 5 above)
Comment (Roms1383) [reply src/types/game_engine.rs]: fixed in 21bf251.

File: src/types/game_engine.rs (hunk 4 above)
Comment (Roms1383) [reply src/types/game_engine.rs]: fixed in 9d938a2.

File: src/types/game_engine.rs (hunk 3 above)
Comment (Roms1383) [reply src/types/game_engine.rs]: renamed in 3308d8c.

File: src/types/game_engine.rs (hunk 2 above)
Comment (Roms1383) [reply src/types/game_engine.rs]: renamed in 3308d8c.

**Pull request merged**